Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -472,7 +472,114 @@ public struct HTTPHeaders: CustomStringConvertible, ExpressibleByDictionaryLiter
+            while let value = parser.nextValue() {
+                values.append(value)
+            }
+        }
+        return values
+    }
+}
+
+private struct HTTPHeaderValueParser {
+    var current: Substring
+
+    init(_ string: String) {
+        self.current = .init(string)
+    }
+
+    mutating func nextValue() -> Substring? {
+        self.popWhitespace()
+        guard !self.current.isEmpty else {
+            return nil
+        }
+        let value: Substring
+        if self.current.first == "\"" {
+            self.pop()
+            guard let nextDoubleQuote = self.nextDoubleQuote() else {
+                return nil
+            }
+            value = self.pop(to: nextDoubleQuote).unescapingQuotes()
+            self.pop()
+            if let comma = self.nextComma() {
+                _ = self.pop(to: comma)
+                self.pop()
+            }
+        } else if let comma = self.nextComma() {
+            value = self.pop(to: comma)
+            self.pop()
+        } else if let whitespace = self.nextWhitespace() {
+            value = self.pop(to: whitespace)
+        } else {
+            value = self.pop(to: self.current.endIndex)
+        }
+        return value
+    }
+
+    private func nextDoubleQuote() -> Substring.Index? {
+        return self.nextDoubleQuote(from: self.current.startIndex)
+    }
+
+    private func nextDoubleQuote(from startIndex: Substring.Index) -> Substring.Index? {
+        guard let nextQuote = self.current[startIndex...].firstIndex(of: "\"") else {
+            return nil
+        }
+        if self.current[self.current.index(before: nextQuote)] == "\\" {
+            // Skip escaped quotes.
+            return self.nextDoubleQuote(from: self.current.index(after: nextQuote))
+        }
+        return nextQuote
+    }
+
+    private func nextComma() -> Substring.Index? {
+        return self.current.firstIndex(of: ",")
+    }
+
+    private mutating func nextWhitespace() -> Substring.Index? {
+        return self.current.firstIndex(where: { $0.isLinearWhitespace })
+    }
+
+    private mutating func popWhitespace() {
+        if let nonWhitespace = self.current.firstIndex(where: { !$0.isLinearWhitespace }) {
+            self.current = self.current[nonWhitespace...]
+        } else {
+            self.current = ""
+        }
+    }
+
+    private mutating func pop() {
+        if self.current.startIndex != self.current.endIndex {
+            self.current = self.current[self.current.index(after: self.current.startIndex)...]
+        }
+    }
+    private mutating func pop(to index: Substring.Index) -> Substring {
+        let value = self.current[..<index]
+        self.current = self.current[index...]
+        return value
+    }
+}
+
+private extension Substring {
+    /// Converts all `\"` to `"`.
+    func unescapingQuotes() -> Substring {
+        return self.split(separator: "\\").reduce("") { (result, part) -> SubSequence in
```

Comment:
We should probably do this lazily to avoid taking an allocation hit here.

Reply:
We can also use `reduce(into:)` which should give us slightly nicer perf characteristics as well.

Reply:
Any pointers on how this could be done lazily?

Reply:
`.lazy` would do the trick I think. ;)